File: src/core/strategies/utils/retry.py
"""Retry logic for LLM API calls (429, 500, 503, timeout)."""

import asyncio
import json
import logging
from collections.abc import Awaitable, Callable
from typing import TypeVar

from src.core.strategies.utils.constants import (
    MAX_RETRIES,
    RETRY_BASE_DELAY,
    RETRYABLE_STATUS_CODES,
)

logger = logging.getLogger(__name__)

T = TypeVar("T")


async def retry_llm_call(
    call_fn: Callable[[], Awaitable[T]],
    *,
    call_description: str,
) -> T:
    """
    Wrap async LLM API call with retry on 429, 500, 503.
    Extracts retry delay from Gemini error when available; else exponential backoff.
    Raises original exception after all retries exhausted.
    """
    last_exc: Exception | None = None
    for attempt in range(MAX_RETRIES + 1):
        try:
            return await call_fn()
        except Exception as e:
            last_exc = e
            if attempt == MAX_RETRIES:
                logger.error(
                    "retry_exhausted | call=%s | attempts=%d | error=%s",
                    call_description,
                    MAX_RETRIES + 1,
                    str(e),
                )
                raise

            status = _extract_status_code(e)
            if status not in RETRYABLE_STATUS_CODES:
                raise

            delay = _extract_retry_delay(e) or (RETRY_BASE_DELAY * (2**attempt))
            logger.warning(
                "retry_scheduled | call=%s | attempt=%d | status=%s | delay_s=%.0f",
                call_description,
                attempt + 1,
                status,
                delay,
            )
            await asyncio.sleep(delay)

    assert last_exc is not None
    raise last_exc


def _extract_status_code(exc: Exception) -> int | None:
    """Extract HTTP status code from exception if present."""
    if hasattr(exc, "status_code"):
        return getattr(exc, "status_code")
    if hasattr(exc, "response") and exc.response is not None:
        return getattr(exc.response, "status_code", None)
    if hasattr(exc, "status"):
        return getattr(exc, "status")
    # Try parsing from message (e.g. "429 Resource Exhausted")
    msg = str(exc).lower()
    for code in RETRYABLE_STATUS_CODES:
        if str(code) in msg:
            return code
    return None


def _extract_retry_delay(exc: Exception) -> float | None:
    """Parse retryDelay from Gemini error JSON when available."""
    if hasattr(exc, "message") and exc.message:
        try:
            # Gemini may return JSON with retryDelay in seconds
            data = json.loads(exc.message)
            if isinstance(data, dict) and "retryDelay" in data:
                return float(data["retryDelay"])
        except (json.JSONDecodeError, TypeError, ValueError):
            pass
    if hasattr(exc, "details") and exc.details:
        try:
            data = json.loads(str(exc.details))
            if isinstance(data, dict) and "retryDelay" in data:
                return float(data["retryDelay"])
        except (json.JSONDecodeError, TypeError, ValueError):
            pass
    return None

File: ts-version/src/core/strategies/utils/retry.ts
import { logger } from '../../../config/logging';
import { MAX_RETRIES, RETRY_BASE_DELAY, RETRYABLE_STATUS_CODES } from './constants';

export async function retryLLMCall<T>(
  callFn: () => Promise<T>,
  callDescription: string
): Promise<T> {
  let lastExc: Error | null = null;

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      return await callFn();
    } catch (e) {
      lastExc = e instanceof Error ? e : new Error(String(e));
      
      if (attempt === MAX_RETRIES) {
        logger.error(
          `retry_exhausted | call=${callDescription} | attempts=${MAX_RETRIES + 1} | error=${String(e)}`
        );
        throw lastExc;
      }

      const status = extractStatusCode(e);
      if (status === null || !RETRYABLE_STATUS_CODES.has(status)) {
        throw e;
      }

      const delay = extractRetryDelay(e) || (RETRY_BASE_DELAY * Math.pow(2, attempt));
      logger.warning(
        `retry_scheduled | call=${callDescription} | attempt=${attempt + 1} | status=${status} | delay_s=${Math.round(delay)}`
      );
      
      await new Promise(resolve => setTimeout(resolve, delay * 1000));
    }
  }

  throw lastExc;
}

function extractStatusCode(exc: any): number | null {
  if (exc && typeof exc === 'object') {
    if ('status_code' in exc) return exc.status_code as number;
    if ('response' in exc && exc.response && 'status_code' in exc.response) {
      return exc.response.status_code as number;
    }
    if ('status' in exc) return exc.status as number;
  }

  // Try parsing from message (e.g. "429 Resource Exhausted")
  const msg = String(exc).toLowerCase();
  for (const code of RETRYABLE_STATUS_CODES) {
    if (msg.includes(code.toString())) {
      return code;
    }
  }
  
  return null;
}

function extractRetryDelay(exc: any): number | null {
  if (exc && typeof exc === 'object') {
    // Check for message property with JSON
    if ('message' in exc && typeof exc.message === 'string') {
      try {
        const data = JSON.parse(exc.message);
        if (typeof data === 'object' && data !== null && 'retryDelay' in data) {
          return Number(data.retryDelay);
        }
      } catch {
        // Ignore JSON parse errors
      }
    }

    // Check for details property with JSON
    if ('details' in exc && typeof exc.details === 'string') {
      try {
        const data = JSON.parse(exc.details);
        if (typeof data === 'object' && data !== null && 'retryDelay' in data) {
          return Number(data.retryDelay);
        }
      } catch {
        // Ignore JSON parse errors
      }
    }
  }

  return null;
}
